cpp/src/arrow/compute/exec/hash_join_node.cc

```diff
@@ -275,30 +275,56 @@ Status HashJoinSchema::ValidateSchemas(JoinType join_type, const Schema& left_sc
 }
 
 std::shared_ptr<Schema> HashJoinSchema::MakeOutputSchema(
-    const std::string& left_field_name_prefix,
-    const std::string& right_field_name_prefix) {
+    const std::string& left_field_name_suffix,
+    const std::string& right_field_name_suffix) {
   std::vector<std::shared_ptr<Field>> fields;
   int left_size = proj_maps[0].num_cols(HashJoinProjection::OUTPUT);
   int right_size = proj_maps[1].num_cols(HashJoinProjection::OUTPUT);
   fields.resize(left_size + right_size);
 
-  for (int i = 0; i < left_size + right_size; ++i) {
-    bool is_left = (i < left_size);
-    int side = (is_left ? 0 : 1);
-    int input_field_id = proj_maps[side]
-                             .map(HashJoinProjection::OUTPUT, HashJoinProjection::INPUT)
-                             .get(is_left ? i : i - left_size);
+  std::unordered_multimap<std::string, int> left_field_map;
```

Comment:
At first I was a little confused why you used a `multimap` instead of a `set`.  Then I remembered that we might have multiple columns with the same name.

However, in order for that case to work, you would need to use `equal_range` instead of `find` down below when you actually search for the field.

In fact, I don't think we have any test cases for this in the hash join tests (and I'm not 100% sure we'd completely support it elsewhere) but it might be an interesting test case.

Either way, can you change the below "finding" code to use `equal_range` instead of `find`?

Reply:
You're making a very important point. One thing, I wanted to do was, to make sure I use the correct order, because I want to bind the column name with the corresponding iterator index. That's the main reason why I selected a `multimap` leaving space for multiple keys too. But again, thinking about this, do we allow duplicate column names in a schema? I might have missed this even though I wrote it like this. 

`equal_range` needs to be there and I will correct it. 

By the way, I also assumed O(m*n) must be heavy if we have columns with 1000s of columns, that's why the code required that map in the first place. Is there a better way to optimize it?

Reply:
Think my comment on multi-column is wrong: multi-columns are supported as I assumed.

Reply:
Updated the PR